// smogon-usage-fetch/dist/esm/parsing/chaos.js
import { toMap, toMapBy } from "lightdash";
/**
 * @private
 */
const mapSpread = (spreadKey) => {
    const [nature, evs] = spreadKey.split(":");
    const [hp, atk, def, spa, spd, spe] = evs.split("/");
    return {
        nature,
        evs: {
            hp: Number(hp),
            atk: Number(atk),
            def: Number(def),
            spa: Number(spa),
            spd: Number(spd),
            spe: Number(spe),
        },
    };
};
/**
 * @private
 */
const mapPokemonData = (rawPokemonData) => {
    return {
        usage: rawPokemonData.usage,
        rawCount: rawPokemonData["Raw count"],
        moves: toMap(rawPokemonData.Moves),
        abilities: toMap(rawPokemonData.Abilities),
        items: toMap(rawPokemonData.Items),
        spreads: toMapBy(rawPokemonData.Spreads, (key) => mapSpread(key), (_key, val) => val),
        happiness: toMapBy(rawPokemonData.Happiness, (key) => Number(key), (_key, val) => val),
        teammates: toMap(rawPokemonData.Teammates),
        checksAndCounters: toMap(rawPokemonData["Checks and Counters"]),
        viabilityCeiling: rawPokemonData["Viability Ceiling"],
    };
};
/**
 * @private
 */
export const mapChaosData = (rawChaosData) => {
    return {
        cutoff: rawChaosData.info.cutoff,
        cutoffDeviation: rawChaosData.info["cutoff deviation"],
        numberOfBattles: rawChaosData.info["number of battles"],
        data: toMapBy(rawChaosData.data, (key) => key, (_key, val) => mapPokemonData(val)),
    };
};
