// smogon-usage-fetch/dist/cjs/parsing/leads.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.leadsFromString = void 0;
const regex_1 = require("./util/regex");
const frequency_1 = require("./util/frequency");
const table_1 = require("./table");
/**
 * Extracts lead data from markdown table.
 *
 * @private
 * @param table Markdown table.
 * @return Lead data items.
 */
const parseLeadTable = (table) => table_1.parseMarkdownTable(table, 5).rows.map((row) => {
    return {
        rank: Number(row[0]),
        pokemon: row[1],
        usagePercentage: frequency_1.convertFrequency(row[2]),
        raw: Number(row[3]),
        rawPercentage: frequency_1.convertFrequency(row[4]),
    };
});
/**
 * @private
 */
const LEADS_TOTAL_REGEX = /Total leads: (-?\d+)/;
/**
 * Parses a smogon leads page.
 *
 * @private
 * @param page Page to parse.
 * @return parsed page.
 */
const leadsFromString = (page) => {
    const rows = page.split("\n");
    const totalRow = rows[0];
    const table = rows.slice(1).join("\n");
    return {
        total: Number(regex_1.getMatchGroup(totalRow, LEADS_TOTAL_REGEX, 1)),
        data: parseLeadTable(table),
    };
};
exports.leadsFromString = leadsFromString;
