// smogon-usage-fetch/dist/cjs/parsing/usages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.usageFromString = void 0;
const regex_1 = require("./util/regex");
const frequency_1 = require("./util/frequency");
const table_1 = require("./table");
/**
 * Extracts usage data from markdown table.
 *
 * @private
 * @param table Markdown table.
 * @return Usage data items.
 */
const parseUsageTable = (table) => table_1.parseMarkdownTable(table, 7).rows.map((row) => {
    return {
        rank: Number(row[0]),
        pokemon: row[1],
        usagePercentage: frequency_1.convertFrequency(row[2]),
        raw: Number(row[3]),
        rawPercentage: frequency_1.convertFrequency(row[4]),
        real: Number(row[5]),
        realPercentage: frequency_1.convertFrequency(row[6]),
    };
});
/**
 * @private
 */
const USAGE_TOTAL_REGEX = /Total battles: (-?\d+)/;
/**
 * @private
 */
const USAGE_WEIGHT_REGEX = /Avg\. weight\/team: (-?[\d.]+)/;
/**
 * Parses a smogon usage page.
 *
 * @private
 * @param page Page to parse.
 * @return parsed page.
 */
const usageFromString = (page) => {
    const rows = page.split("\n");
    const totalRow = rows[0];
    const weightRow = rows[1];
    const table = rows.slice(2).join("\n");
    return {
        total: Number(regex_1.getMatchGroup(totalRow, USAGE_TOTAL_REGEX, 1)),
        weight: Number(regex_1.getMatchGroup(weightRow, USAGE_WEIGHT_REGEX, 1)),
        data: parseUsageTable(table),
    };
};
exports.usageFromString = usageFromString;
